Default valued generic proc parameter causes missing parameter in generated C code.

### Description

Code:
```nim
type
  Item*[T] = object
    pos*: int

proc initItem[T](p:int=10000) : Item[T] = 
    result = Item[T](p)

proc doit[T](x:Item[T], s:Item[T]=initItem) : string = 
    return $x.pos

let x = Item[string](pos:100)
echo doit(x)
```



### Nim Version

Nim Compiler Version 1.6.8 [Linux: amd64]
Compiled at 2022-10-05
Copyright (c) 2006-2021 by Andreas Rumpf

git hash: c9f46ca8c9eeca8b5f68591b1abe14b962f80a4c
active boot switches: -d:release


### Current Output

```text
Output:

vagrant@devvm:~/nimly$ nim c -r tests/nim_bug_repro.nim 
Hint: used config file '/home/vagrant/Nim/config/nim.cfg' [Conf]
Hint: used config file '/home/vagrant/Nim/config/config.nims' [Conf]
Hint: used config file '/home/vagrant/nimly/tests/config.nims' [Conf]
..................................................................
CC: ../../Nim/lib/system/io.nim
CC: ../../Nim/lib/system.nim
CC: nim_bug_repro.nim
/home/vagrant/.cache/nim/nim_bug_repro_d/@mnim_bug_repro.nim.c: In function ‘NimMainModule’:
/home/vagrant/.cache/nim/nim_bug_repro_d/@mnim_bug_repro.nim.c:169:62: error: expected expression before ‘)’ token
  169 |  T1_[0] = dollar___nim95bug95repro_25(x__nim95bug95repro_11, );
      |                                                              ^
Error: execution of an external compiler program 'gcc -c  -w -fmax-errors=3   -I/home/vagrant/Nim/lib -I/home/vagrant/nimly/tests -o /home/vagrant/.cache/nim/nim_bug_repro_d/@mnim_bug_repro.nim.c.o /home/vagrant/.cache/nim/nim_bug_repro_d/@mnim_bug_repro.nim.c' failed with exit code: 1
```

### Expected Output



I'd expect `proc doit[T](x:Item[T], s:HashSet[T]=initHashSet) : string =` to be either valid code or I should get a compile error. 

I do get an error, but it shows up as a gcc error, complaining about a missing parameter in a call. 

This feels very opaque and doesn't show me what is wrong. My expectation would have been for the Nim compiler to error out, not for it to show up as a gcc error.



### Possible Solution

Could Nim compiler error out instead of gcc crapping out on generated code? 

### Additional Information

This came up in a project where I defined `$` on a generic type, with a generic second parameter as HashSet[T]=initHashSet. In a unit test I got the error above. So it's not an academic question ;)

--- a/compiler/semcall.nim
+++ b/compiler/semcall.nim
@@ -595,7 +595,7 @@ proc inferWithMetatype(c: PContext, formal: PType,
     result = copyTree(arg)
     result.typ = formal
 
-proc updateDefaultParams(call: PNode) =
+proc updateDefaultParams(c: PContext, call: PNode) =
   # In generic procs, the default parameter may be unique for each
   # instantiation (see tlateboundgenericparams).
   # After a call is resolved, we need to re-assign any default value
@@ -605,8 +605,18 @@ proc updateDefaultParams(call: PNode) =
   let calleeParams = call[0].sym.typ.n
   for i in 1..<call.len:
     if nfDefaultParam in call[i].flags:
-      let def = calleeParams[i].sym.ast
+      let formal = calleeParams[i].sym
+      let def = formal.ast
       if nfDefaultRefsParam in def.flags: call.flags.incl nfDefaultRefsParam
+      # mirrored with sigmatch:
+      if def.kind == nkEmpty:
+        # The default param value is set to empty in `instantiateProcType`
+        # when the type of the default expression doesn't match the type
+        # of the instantiated proc param:
+        pushInfoContext(c.config, call.info, call[0].sym.detailedInfo)
+        typeMismatch(c.config, def.info, formal.typ, def.typ, formal.ast)
+        popInfoContext(c.config)
+        def.typ = errorType(c)
       call[i] = def
 
 proc getCallLineInfo(n: PNode): TLineInfo =
@@ -727,7 +737,7 @@ proc semResolvedCall(c: PContext, x: var TCandidate,
   result[0] = newSymNode(finalCallee, getCallLineInfo(result[0]))
   if finalCallee.magic notin {mArrGet, mArrPut}:
     result.typ = finalCallee.typ.returnType
-  updateDefaultParams(result)
+  updateDefaultParams(c, result)
 
 proc canDeref(n: PNode): bool {.inline.} =
   result = n.len >= 2 and (let t = n[1].typ;

--- a/compiler/seminst.nim
+++ b/compiler/seminst.nim
@@ -292,6 +292,8 @@ proc instantiateProcType(c: PContext, pt: TypeMapping,
         # the user calls an explicit instantiation of the proc (this is
         # the only way the default value might be inserted).
         param.ast = errorNode(c, def)
+        # we know the node is empty, we need the actual type for error message
+        param.ast.typ = def.typ
       else:
         param.ast = fitNodePostMatch(c, typeToFit, converted)
       param.typ = result[i]

--- a/compiler/seminst.nim
+++ b/compiler/seminst.nim
@@ -253,9 +253,13 @@ proc instantiateProcType(c: PContext, pt: TypeMapping,
     var typeToFit = resulti
 
     let needsStaticSkipping = resulti.kind == tyFromExpr
+    let needsTypeDescSkipping = resulti.kind == tyTypeDesc and tfUnresolved in resulti.flags
     result[i] = replaceTypeVarsT(cl, resulti)
     if needsStaticSkipping:
       result[i] = result[i].skipTypes({tyStatic})
+    if needsTypeDescSkipping:
+      result[i] = result[i].skipTypes({tyTypeDesc})
+      typeToFit = result[i]
 
     # ...otherwise, we use the instantiated type in `fitNode`
     if (typeToFit.kind != tyTypeDesc or typeToFit.base.kind != tyNone) and

--- a/compiler/sigmatch.nim
+++ b/compiler/sigmatch.nim
@@ -2795,7 +2795,8 @@ proc matches*(c: PContext, n, nOrig: PNode, m: var TCandidate) =
           # The default param value is set to empty in `instantiateProcType`
           # when the type of the default expression doesn't match the type
           # of the instantiated proc param:
-          pushInfoContext(c.config, m.call.info, m.calleeSym.detailedInfo)
+          pushInfoContext(c.config, m.call.info,
+            if m.calleeSym != nil: m.calleeSym.detailedInfo else: "")
           typeMismatch(c.config, formal.ast.info, formal.typ, formal.ast.typ, formal.ast)
           popInfoContext(c.config)
           formal.ast.typ = errorType(c)

--- a/compiler/sigmatch.nim
+++ b/compiler/sigmatch.nim
@@ -2790,14 +2790,15 @@ proc matches*(c: PContext, n, nOrig: PNode, m: var TCandidate) =
           m.firstMismatch.formal = formal
           break
       else:
+        # mirrored with updateDefaultParams:
         if formal.ast.kind == nkEmpty:
           # The default param value is set to empty in `instantiateProcType`
           # when the type of the default expression doesn't match the type
           # of the instantiated proc param:
-          localError(c.config, m.call.info,
-                     ("The default parameter '$1' has incompatible type " &
-                      "with the explicitly requested proc instantiation") %
-                      formal.name.s)
+          pushInfoContext(c.config, m.call.info, m.calleeSym.detailedInfo)
+          typeMismatch(c.config, formal.ast.info, formal.typ, formal.ast.typ, formal.ast)
+          popInfoContext(c.config)
+          formal.ast.typ = errorType(c)
         if nfDefaultRefsParam in formal.ast.flags:
           m.call.flags.incl nfDefaultRefsParam
         var defaultValue = copyTree(formal.ast)

--- a/tests/proc/twrongdefaultvalue.nim
+++ b/tests/proc/twrongdefaultvalue.nim
@@ -1,0 +1,25 @@
+discard """
+  cmd: "nim check $file"
+  action: reject
+  nimout: '''
+twrongdefaultvalue.nim(20, 12) template/generic instantiation of `doit` from here
+twrongdefaultvalue.nim(17, 37) Error: type mismatch: got <proc (p: int): Item[initItem.T]> but expected 'Item[system.string]'
+twrongdefaultvalue.nim(25, 3) template/generic instantiation of `foo` from here
+twrongdefaultvalue.nim(23, 33) Error: type mismatch: got <string> but expected 'int'
+'''
+"""
+
+block: # issue #21258
+  type Item[T] = object
+    pos: int
+  proc initItem[T](p:int=10000) : Item[T] = 
+    result = Item[T](p)
+  proc doit[T](x:Item[T], s:Item[T]=initItem) : string = 
+    return $x.pos
+  let x = Item[string](pos:100)
+  echo doit(x)
+
+block: # issue #21258, reduced case
+  proc foo[T](x: seq[T], y: T = "foo") =
+    discard
+  foo @[1, 2, 3]

